site

```diff
@@ -0,0 +1,175 @@
+#!/bin/bash
+
+set -eu
+
+function check_version() {
+    local v
+
+    for v in "${VERSION[@]}"; do
+        [[ ${v} -ne ${1-0} ]] && break
+        unset v
+        shift
+    done
+
+    [[ ${v-0} -ge ${1-0} ]]
+}
+
+function template() {
+    local file="${SITE_HOME}/${1}"
+    local mode="${2}"
+    local content="${3}"
+
+    sed -n 's/^\s*|//p' <<<"${content}" >"${file}"
+    chmod "${mode}" "${file}"
+}
+
+function run() {
+    "${SITE_HOME}/site_wrapper" "${@}"
+}
+
+function cmd_show() {
+    find "${SITE_USER}" -mindepth 1 -maxdepth 1 -type d |
+        xargs -i{} bash -c '
+            if [[ -x "{}/site_wrapper" ]]; then
+                echo "$(basename "{}") $(grep -o "EXTRA_VARS.*" "{}/site_wrapper")"
+            fi
+        '
+}
+
+function cmd_yamllint() {
+    yamllint -c .yamllint .
+}
+
+function cmd_statedump() {
+    run make nodes.statedump
+}
+
+function cmd_shell() {
+    run bash -i
+}
+
+function cmd_create() {
+    local opts
+
+    opts=(
+        "--symlinks"
+        "--clear"
+    )
+    check_version 3 6 && opts+=("--prompt" "SITE:${SITE_HOME}")
+    check_version 3 9 && opts+=("--upgrade-deps")
+
+    ${PYTHON} -m venv "${opts[@]}" "${SITE_HOME}"
+    template "site_wrapper" 0755 "
+        |#!/bin/bash
+        |
+        |export SITE_HOME=\"\$(realpath \"\$(dirname \"\${0}\")\")\"
+        |export SITE_NAME=\"${SITE_NAME}\"
+        |export SITE_ROOT=\"${SITE_ROOT}\"
+        |
+        |export EXTRA_VARS=\"${EXTRA_VARS-}\"
+        |
+        |source \"\${SITE_HOME}/bin/activate\"
+        |
+        |cd \"\${SITE_HOME}\"
+        |
+        |exec \"\${@}\"
+    "
+
+    template "Makefile" 0644 "
+        |PHONY: \$(MAKECMDGOALS) all
+        |
+        |\$(or \$(MAKECMDGOALS), all):
+        |	@make -C \"\${SITE_ROOT}/playbooks\" \$(MAKECMDGOALS)
+    "
+
+    cp "${SITE_ROOT}/requirements.txt" "${SITE_HOME}"
+
+    run pip install --upgrade -r requirements.txt
+}
+
+function cmd_build() {
+    run make
+}
+
+function cmd_clean() {
+    run make clean
+}
+
+function cmd_destroy() {
+    cmd_clean
+    rm -rf "${SITE_HOME}"
+}
+
+SITE_USER="$(realpath "${HOME}/.site")"
+
+mkdir -p "${SITE_USER}"
+
+CMD="${1-show}"
+
+if [[ "${CMD}" == "show" ]]; then
+    cmd_show
+    exit 0
+elif [[ "${CMD}" == "yamllint" ]]; then
+    cmd_yamllint
+    exit 0
+fi
+
+PYTHON="$(echo -n "${PATH}" |
```

Comment:
just wondering if we can use something like $which python |  xargs -i{} bash -c '{} --version' instead of listing all the PATHs

Reply:
I got quite puzzled when I tried to execute make on this branch:
```
shwethaacharya:sit-environment$ EXTRA_VARS="backend=xfs" make
Python 2.7.18
Python 2.7.18
Python 3.4+ is required
make: *** [Makefile:4: test] Error 1
shwethaacharya:sit-environment$ 
```

Basically python2 package versions are getting formatted differently than python3, 
```
shwethaacharya:sit-environment$ echo -n "${PATH}" |tr ':' '\n' |xargs -i{} find {} -maxdepth 1 -type f -regex '.*/python[0-9.]*' 2>/dev/null | xargs -i{} bash -c "echo \$({} --version |  tr -dc '0-9.') {}"|LC_ALL=C sort -Vr
Python 2.7.18
/usr/bin/python2.7
3.13.02 /usr/bin/python3.13
3.12.0 /usr/bin/python3.12
3.11.6 /usr/bin/python3.11
3.10.13 /usr/bin/python3.10
3.9.18 /usr/bin/python3.9
3.8.18 /usr/bin/python3.8
3.7.17 /usr/bin/python3.7
3.6.15 /usr/bin/python3.6
shwethaacharya:sit-environment$
```

and hence, a wrong python version appears for the command `head -n 1`:
```
shwethaacharya:sit-environment$ PYTHON="$(echo -n "${PATH}" |tr ':' '\n' |xargs -i{} find {} -maxdepth 1 -type f -regex '.*/python[0-9.]*' 2>/dev/null |xargs -i{} bash -c "echo \$({} --version | tr -dc '0-9.') {}" |LC_ALL=C sort -Vr |head -n 1 |cut -d" " -f 2-)"
Python 2.7.18
shwethaacharya:sit-environment$
```
I am not sure if we want to handle this scenario with python2 or any future such possible occurences or should ignore it considering python2 is EOLed.

Just adding my observation here.

Reply:
That's interesting. I do want to handle those cases for sure. The problem is that the format is the expected one, so this shouldn't happen.

I've tested what happens and it seems that python 2.x shows the version info in the stderr instead of the stdout, so the version parsed is actually an empty string.

I'll update it. Thanks for catching it !